### PR TITLE
Vocal and Textual Navigation Assistance

### PR DESCRIPTION
**Pull request by Rxann**
Uses the generate directions native to play audio and show popups with directions. this took me way longer than it should have.

### CODE REVIEW

File: src/backend/looped/vehicle/navigation.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		{
+			g_notification_service->push_success("Voice Navigation", "Voice Navigation has been enabled");
+		}
+
+		virtual void on_tick() override
+		{
+			if (g.self.nav_voice)
+			{
+				if (g_player_service->get_self()->get_current_vehicle() != nullptr)
+					g_fiber_pool->queue_job([] {
```
Comment (gir489returns) [inline src/backend/looped/vehicle/navigation.cpp]: This is stupid, why are you queueing a job from within a tick, which runs inside a script thread?

Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp]: Because I call a native and audio functions.

File: src/backend/looped/vehicle/navigation.cpp [hunk 2]
```diff
@@ -22,27 +31,270 @@ namespace big
...
+		3 = In (distToNxJunction) Turn Left
+		4 = In (distToNxJunction) Turn Right
+		5 = In (distToNxJunction) Keep Straight
+		6 = In (distToNxJunction) Turn Sharply To The Left
+		7 = In (distToNxJunction) Turn Sharply To The Right
+		8 = Route is being recalculated or the navmesh is confusing. This happens randomly during the drive but consistently at {2044.0358, 2996.6116, 44.9717} if you face towards the bar and the route needs you to turn right. In that particular case, it could be a bug with how the turn appears to be 270 deg. CCW instead of "right." Either way, this seems to be the engine saying "I don't know the route right now."
+			*/
+
+
+			g_fiber_pool->queue_job([] {
```
Comment (gir489returns) [inline src/backend/looped/vehicle/navigation.cpp:32]: Remove this.

File: src/backend/looped/vehicle/navigation.cpp [hunk 3]
```diff
@@ -22,27 +31,270 @@ namespace big
...
+				float dist;
+				float v;
+				// most important variables have been declared upfront
+				blip::get_blip_location(waypoint, (int)BlipIcons::Waypoint);
+				bool doesPointExist = blip::get_blip_location(waypoint, (int)BlipIcons::Waypoint);
+				// the if statement allows the function to not run if there is no player waypoint
+				if (doesPointExist)
+				{
+					PATHFIND::GENERATE_DIRECTIONS_TO_COORD(waypoint.x, waypoint.y, waypoint.z, 1, &direction, &v, &dist);
+					feetToAction = math::divided_by_ten(math::meters_to_feet(dist));
```
Comment (gir489returns) [inline src/backend/looped/vehicle/navigation.cpp:45]: Seriously? You need a helper function to divide by ten?

File: src/file_manager.hpp [hunk 4]
```diff
@@ -28,7 +28,7 @@ namespace big
 
 	private:
 		std::filesystem::path m_base_dir;
-
+		std::filesystem::path m_nav_dir;
```
Comment (tupoy-ya) [inline src/file_manager.hpp:31]: What is this?

File: src/backend/looped/vehicle/navigation.cpp (hunk 1 above)
Comment (xiaoxiao921) [reply src/backend/looped/vehicle/navigation.cpp]: https://github.com/YimMenu/YimMenu/blob/44f79a428495bf4b63a6e905a5574e6cc3011a36/src/backend/backend.cpp#L42

looped commands already run inside a script, you don't need `queue_job`

File: src/backend/looped/vehicle/navigation.cpp [hunk 5]
```diff
@@ -22,27 +31,270 @@ namespace big
...
+							}
+							break;
+						}
+						//g_notification_service->push_warning("Navigation", std::format("In {} feet, keep right", feetToAction));
+						break;
+					case 5:
+						switch (feetToAction)
+						{
+						case 130:
+							if (g.self.nav_voice)
```
Comment (gir489returns) [inline src/backend/looped/vehicle/navigation.cpp:143]: Why are you checking g.self.nav_voice a million times? This should either be checked at the head or returned if the user has this off.

File: src/file_manager.cpp [hunk 6]
```diff
@@ -2,54 +2,58 @@
...
-    }
-
-    const std::filesystem::path& file_manager::get_base_dir()
-    {
-        return m_base_dir;
-    }
-    
+	bool file_manager::init(const std::filesystem::path& base_dir)
+	{
+		m_base_dir       = base_dir;
```
Comment (gir489returns) [inline src/file_manager.cpp:7]: Revert these changes, you are not supposed to touch this class.

File: src/file_manager.hpp [hunk 7]
```diff
@@ -8,11 +8,11 @@ namespace big
 	class file_manager final
 	{
 	public:
-		file_manager() = default;
-		virtual ~file_manager() = default;
-		file_manager(const file_manager&) = delete;
-		file_manager(file_manager&&) noexcept = delete;
-		file_manager& operator=(const file_manager&) = delete;
+		file_manager()                                   = default;
```
Comment (gir489returns) [inline src/file_manager.hpp:11]: Revert these too.

File: src/hooking.cpp [hunk 8]
```diff
@@ -56,7 +56,8 @@ namespace big
 		detour_hook_helper::add<hooks::get_network_event_data>("GNED", g_pointers->m_gta.m_get_network_event_data);
 		detour_hook_helper::add<hooks::write_player_gamer_data_node>("WPGDN", g_pointers->m_gta.m_write_player_gamer_data_node);
 
-		detour_hook_helper::add<hooks::invalid_mods_crash_detour>("IMCD", g_pointers->m_gta.m_invalid_mods_crash_detour);
+		// removed due to issues with audio.
```
Comment (BugisoftRSG) [inline src/hooking.cpp:59]: Can you please provide some informations on what is happening if this is active

File: src/backend/looped/vehicle/navigation.cpp (hunk 1 above)
Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp]: Ok, I will change this. In main.cpp do i need to add it to the script manager?

File: src/util/audio.hpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
+#pragma once
+#include <Windows.h>
+#include <filesystem>
+#include <mmsystem.h>
+
+#pragma comment(lib, "winmm.lib")
+
+namespace big::audio
+{
+	enum string_code
```
Comment (gir489returns) [inline src/util/audio.hpp:10]: This belongs in either gta/enums.hpp or core/enums.hpp, it depends if the game uses them or not, I have no idea where they come from.

File: src/views/self/view_self.cpp [hunk 10]
```diff
@@ -163,6 +167,26 @@ namespace big
 				ImGui::EndCombo();
 			}
 		});
+		ImGui::NewLine();
+		ImGui::Checkbox("Navigation", &g.self.nav);
+		components::options_modal("Navigation", [] {
```
Comment (gir489returns) [inline src/views/self/view_self.cpp:172]: These should belong in the Vehicle menu, not self, as they are not relevant here.

File: src/core/settings.hpp [hunk 11]
```diff
@@ -336,6 +336,10 @@ namespace big
 			bool custom_weapon_stop    = true;
 			std::string persist_outfit = "";
 			bool persist_outfits_mis   = false;
+			bool nav                   = false;
```
Comment (tupoy-ya) [inline src/core/settings.hpp:339]: You could create a nav struct, something like:
```cpp
struct nav
{
	enabled = false;
	voice      = false;
	text        = false;
	NLOHMANN_DEFINE_TYPE_INTRUSIVE(nav, enabled, voice, text)
} nav {};
```

File: src/views/self/view_self.cpp [hunk 12]
```diff
@@ -163,6 +167,26 @@ namespace big
...
+			ImGui::BeginDisabled(!g.self.nav);
+			ImGui::NewLine();
```
Comment (gir489returns) [inline src/views/self/view_self.cpp:178]: These should belong in the Vehicle menu, not self, as they are not relevant here.

File: src/views/self/view_self.cpp (hunk 10 above)
Comment (gir489returns) [inline src/views/self/view_self.cpp:172]: These should belong in the Vehicle menu, not self, as they are not relevant here.

File: src/views/self/view_self.cpp (hunk 12 above)
Comment (gir489returns) [inline src/views/self/view_self.cpp:178]: These should belong in the Vehicle menu, not self, as they are not relevant here.

**Review by gir489returns — CHANGES_REQUESTED** [review]

File: src/hooking.cpp (hunk 8 above)
Comment (Rxann) [reply src/hooking.cpp:59]: Yes, When I start driving, there is an exception produced. It crashes the game. I can provide the exception that is printed out later if needed.

File: CMakeLists.txt [hunk 13]
```diff
@@ -54,7 +54,7 @@ target_include_directories(YimMenu PRIVATE
 )
 
 target_precompile_headers(YimMenu PRIVATE "${SRC_DIR}/common.hpp")
-target_link_libraries(YimMenu PRIVATE pugixml minhook AsyncLogger dbghelp imgui cpr Zydis lua_static)
+target_link_libraries(YimMenu PRIVATE pugixml minhook AsyncLogger dbghelp imgui cpr Zydis lua_static winmm.lib)
```
Comment (Yimura) [inline CMakeLists.txt:57]: You can omit the `.lib`.

File: src/core/settings.hpp [hunk 14]
```diff
@@ -355,7 +359,8 @@ namespace big
 			// do not save below entries
 			bool dance_mode = false;
 
-			NLOHMANN_DEFINE_TYPE_INTRUSIVE(self, ipls, ptfx_effects, clean_player, force_wanted_level, free_cam, invisibility, local_visibility, never_wanted, no_ragdoll, noclip, off_radar, super_run, no_collision, unlimited_oxygen, no_water_collision, wanted_level, god_mode, part_water, proof_bullet, proof_fire, proof_collision, proof_melee, proof_explosion, proof_steam, proof_drown, proof_water, proof_mask, mobile_radio, fast_respawn, auto_tp, super_jump, beast_jump, healthregen, healthregenrate, hud, superman, custom_weapon_stop, persist_outfit, persist_outfits_mis)
+
```
Comment (Yimura) [inline src/core/settings.hpp:362]: Extra newline

File: src/hooking.cpp (hunk 8 above)
Comment (gir489returns) [reply src/hooking.cpp:59]: You have to fix it then instead of just turning it off. That's not a solution.

File: src/backend/looped/vehicle/navigation.cpp [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+				bool doesPointExist = blip::get_blip_location(waypoint, (int)BlipIcons::Waypoint);
+
+				if (doesPointExist)
+				{
+					PATHFIND::GENERATE_DIRECTIONS_TO_COORD(waypoint.x, waypoint.y, waypoint.z, 1, &direction, &v, &dist);
+					feetToAction = math::divided_by_ten(math::meters_to_feet(dist));
+					switch (direction)
+					{
+					case 1:
+						if (g.self.nav_text)
```
Comment (tupoy-ya) [inline src/backend/looped/vehicle/navigation.cpp:49]: Just make a helper function that will check both `nav_text` or `nav_voice` why do it every time?

File: src/backend/looped/vehicle/navigation.cpp [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+		using looped_command::looped_command;
+
+		virtual void on_enable() override
+		{
+			g_notification_service->push_success("Navigation", "Navigation has been enabled");
+		}
+
+		virtual void on_tick() override
+		{
+			g_fiber_pool->queue_job([] {
```
Comment (tupoy-ya) [inline src/backend/looped/vehicle/navigation.cpp:32]: No need to do fiber pool here, `on_tick()` is already run in fiber pool.

File: src/file_manager.cpp (hunk 6 above)
Comment (Yimura) [reply src/file_manager.cpp:7]: Yep leave this class alone, see a later comment on how to properly make use of the file_manager.

File: src/backend/looped/vehicle/navigation.cpp (hunk 3 above)
Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp:45]: It is to make it more streamlined.  Plus it is inlined so it is pretty much doing the same thing but putting it in another place

File: src/util/audio.hpp [hunk 17]
```diff
@@ -0,0 +1,145 @@
+#pragma once
+#include <Windows.h>
+#include <filesystem>
+#include <mmsystem.h>
+
+#pragma comment(lib, "winmm.lib")
```
Comment (tupoy-ya) [inline src/util/audio.hpp:6]: Isn't it already included in CMakeLists.txt?

File: src/util/audio.hpp [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+namespace big::audio
+{
+	enum string_code
```
Comment (tupoy-ya) [inline src/util/audio.hpp:10]: According to `CONTRIBUTING.md` you should use UpperCamelCase for enum names
and ALL_UPPER for enum members.

File: src/file_manager.hpp (hunk 4 above)
Comment (Yimura) [inline src/file_manager.hpp:31]: remove

File: src/util/audio.hpp (hunk 17 above)
Comment (Yimura) [inline src/util/audio.hpp:6]: This is handled by CMakeLists.txt, remove this from the source code.

File: src/util/audio.hpp [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+		eFeetUntilAction,
+		eKeepLeft,
+		eKeepRight,
+		eStraight,
+		eTurnLeft,
+		eTurnRight,
+		eFiveHundredFeet,
+		eError
+	};
+	string_code hash_input(std::string const& input)
```
Comment (tupoy-ya) [inline src/util/audio.hpp:23]: Does anyone see a hash here?
Maybe use a joaat or something.

File: src/util/audio.hpp
Comment (Yimura) [inline src/util/audio.hpp:1]: Whole file could be moved to a service, tbh.

File: src/util/audio.hpp [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+	{
+		if (input == "RouteRecalc")
+		{
+			return eRecalc;
+		}
+		else if (input == "HighlightedRoute")
```
Comment (tupoy-ya) [inline src/util/audio.hpp:29]: How about using `case` instead?

File: src/util/audio.hpp [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+		{
+			return eError;
+		}
+	}
+
+	inline void play_sound_nav(std::string audio_item, int feet)
+	{
+		std::filesystem::path base = std::getenv("appdata");
+		base /= "YimMenu";
+		base /= "navigation";
```
Comment (tupoy-ya) [inline src/util/audio.hpp:71]: You should use `file_manager` here.

File: src/hooking.cpp (hunk 8 above)
Comment (Rxann) [reply src/hooking.cpp:59]: It's a solution because it fixed the problem. But l, I can't find a good way to try to fix it due to the fact that the exception printed makes no sense(to me atleast).

Comment (gir489returns) [reply src/hooking.cpp:59]: It fixed YOUR problem. Learn the difference.

File: src/util/math.hpp [hunk 22]
```diff
@@ -9,6 +9,17 @@ namespace big::math
 		return (float)radian;
 	}
 
+	inline int meters_to_feet(int meters)
```
Comment (tupoy-ya) [inline src/util/math.hpp:12]: Doesn't `mps_to_speed` or `speed_to_mps` already do this?
Even if it doesn't i would rather add it there instead of adding it here.
Also why use an `int` instead of a `float`?

File: src/util/audio.hpp (hunk 21 above)
Comment (Yimura) [inline src/util/audio.hpp:71]: Use file_manager:

```cpp
#include "file_manager.hpp"

folder navigation_folder = g_file_manager.get_project_folder("./navigation/");
```

File: src/views/self/view_self.cpp [hunk 23]
```diff
@@ -7,10 +7,14 @@
 #include "util/scripts.hpp"
 #include "views/view.hpp"
 
+#include <fiber_pool.hpp>
 #include <script/globals/GPBD_FM_3.hpp>
+#include <thread_pool.hpp>
```
Comment (tupoy-ya) [inline src/views/self/view_self.cpp:12]: Why include it here?

File: src/views/self/view_self.cpp [hunk 24]
```diff
@@ -163,6 +167,30 @@ namespace big
 				ImGui::EndCombo();
 			}
 		});
+		ImGui::NewLine();
```
Comment (tupoy-ya) [inline src/views/self/view_self.cpp:170]: What does `ImGui::NewLine();` even do?
ImGui does adds elements an the next line automatically so why do this?

File: src/backend/looped/vehicle/navigation.cpp (hunk 5 above)
Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp:143]: To know if the sound is enabled to play the sound if needed...

File: src/views/self/view_self.cpp [hunk 25]
```diff
@@ -163,6 +167,30 @@ namespace big
...
+			ImGui::NewLine();
+
+			ImGui::Checkbox("Vocal Navigation", &g.self.nav_voice);
+			if (ImGui::IsItemHovered())
+				ImGui::SetTooltip("Vocal Navigation is not perfected yet. It gets you there...");
+			ImGui::Checkbox("Textual Navigation", &g.self.nav_text);
+			if (ImGui::IsItemHovered())
+				ImGui::SetTooltip("Provides Popups with directions.");
+			ImGui::NewLine();
+			ImGui::Text("Make sure you have the following .wav files installed to the navigation directory in YimMenu.");
```
Comment (tupoy-ya) [inline src/views/self/view_self.cpp:189]: Source?

File: src/util/audio.hpp [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+		case eRecalc:
+			base /= "routerecalc.WAV";
+			PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+			break;
+
+		case eFeetUntilAction:
+			switch (feet)
+			{
+			case 5280:
+				base /= "mile.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:83]: Can be replaced with:

```cpp
navigation_folder.get_file("./mile.wav");
```

File: src/util/audio.hpp [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+				PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+				break;
+			case 1000:
+				base /= "thousandfeet.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:87]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/util/audio.hpp [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+				PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+				break;
+
+			case 500:
+				base /= "fivehundredfeet.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:92]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/util/audio.hpp [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+				PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+				break;
+			case 100:
+				base /= "onehundredfeet.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:96]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/util/audio.hpp [hunk 30]
```diff
@@ -0,0 +1,145 @@
...
+				PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+				break;
+			}
+			break;
+		case eFiveHundredFeet:
+			base /= "fivehundredfeet.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:102]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/hooking.cpp (hunk 8 above)
Comment (Rxann) [reply src/hooking.cpp:59]: I know the difference, but as far as I know the invalid mods detour hasn't exactly been used a lot.

File: src/util/audio.hpp [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+			PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+			break;
+
+		case eKeepLeft:
+			base /= "keepleft.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:107]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/file_manager.cpp [hunk 32]
```diff
@@ -2,54 +2,58 @@
...
+		auto edited_base = base_dir;
+		edited_base /= "navigation";
```
Comment (tupoy-ya) [inline src/file_manager.cpp:9]: Don't touch that.

File: src/util/audio.hpp [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+			PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+			break;
+
+		case eKeepRight:
+			base /= "keepright.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:112]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/util/audio.hpp [hunk 34]
```diff
@@ -0,0 +1,145 @@
...
+			PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+			break;
+		case eStraight:
+			base /= "straight.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:116]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/util/audio.hpp [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+			PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+			break;
+		case eTurnLeft:
+			base /= "turnleft.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:120]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/util/audio.hpp [hunk 36]
```diff
@@ -0,0 +1,145 @@
...
+			PlaySound(TEXT(base.string().c_str()), NULL, SND_FILENAME | SND_ASYNC);
+			break;
+
+		case eTurnRight:
+			base /= "turnright.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:125]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: src/util/audio.hpp [hunk 37]
```diff
@@ -0,0 +1,145 @@
...
+			PlaySound(TEXT(base.c_str()), NULL, SND_FILENAME | SND_ASYNC);
+			return true;
+		}
+		catch
+		{
+			return false;
+		}
+	}
+
+}
```
Comment (Yimura) [inline src/util/audio.hpp:145]: You have this helper function but don't even use it? Every other place you use `::PlaySound()`.

**Review by tupoy-ya — CHANGES_REQUESTED** [review]
Maybe missed something or duplicated what others have said.

File: src/util/audio.hpp (hunk 9 above)
Comment (Rxann) [reply src/util/audio.hpp:10]: Core

Comment (Rxann) [reply src/util/audio.hpp:10]: and I made them

File: src/hooking.cpp (hunk 8 above)
Comment (BugisoftRSG) [reply src/hooking.cpp:59]: Can't wait for some modders to drive around with invalid vehicle mods and crash once coming nearby.

File: src/util/math.hpp [hunk 38]
```diff
@@ -9,6 +9,17 @@ namespace big::math
...
+	{
+		float conversion_rate = 3.28084;
+		return (int)(meters * conversion_rate);
+	}
+
+	inline int divided_by_ten(int feet)
+	{
+		return (int)(feet / 10);
+	}
```
Comment (Yimura) [inline src/util/math.hpp:21]: Seems a bit overkill to move divide by 10 here.

Should basically be coded right in where you need it.

File: src/views/self/view_self.cpp (hunk 23 above)
Comment (Yimura) [inline src/views/self/view_self.cpp:12]: These are includes from YimMenu source code, use `#include ""` instead of `#include <>`

File: src/views/self/view_self.cpp [hunk 39]
```diff
@@ -7,10 +7,14 @@
 #include "util/scripts.hpp"
 #include "views/view.hpp"
 
+#include <fiber_pool.hpp>
```
Comment (Yimura) [inline src/views/self/view_self.cpp:10]: These are includes from YimMenu source code, use `#include ""` instead of `#include <>`

File: src/file_manager.hpp (hunk 4 above)
Comment (Rxann) [reply src/file_manager.hpp:31]: It was to ensure the navigation directory was there, where the sounds that are used should be placed.

File: src/core/settings.hpp (hunk 11 above)
Comment (Rxann) [reply src/core/settings.hpp:339]: Good Idea, I will implement this

File: src/util/audio.hpp [hunk 40]
```diff
@@ -0,0 +1,145 @@
...
+
+	inline void play_sound_nav(std::string audio_item, int feet)
+	{
+		std::filesystem::path base = std::getenv("appdata");
+		base /= "YimMenu";
+		base /= "navigation";
+		switch (hash_input(audio_item))
+		{
+		case eRecalc:
+			base /= "routerecalc.WAV";
```
Comment (Yimura) [inline src/util/audio.hpp:75]: Can be replaced with:

```cpp
navigation_folder.get_file("./XXX.wav");
```

File: CMakeLists.txt (hunk 13 above)
Comment (tupoy-ya) [inline CMakeLists.txt:57]: You should remove the .lib extension here.

Comment (Rxann) [reply CMakeLists.txt:57]: My bad, I thought I needed it

File: src/backend/looped/vehicle/navigation.cpp (hunk 15 above)
Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp:49]: so just do like if(checkvoice()) or something?

File: src/backend/looped/vehicle/navigation.cpp (hunk 16 above)
Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp:32]: Thanks

File: src/views/self/view_self.cpp (hunk 24 above)
Comment (gir489returns) [reply src/views/self/view_self.cpp:170]: Undo a SameLine() or force a new line when in a horizontal-layout context. Should probably be removed.

File: src/backend/looped/vehicle/navigation.cpp (hunk 5 above)
Comment (xiaoxiao921) [reply src/backend/looped/vehicle/navigation.cpp:143]: did you even read what gir said or are you trolling?

File: src/backend/looped/vehicle/navigation.cpp
Comment (Yimura) [inline src/backend/looped/vehicle/navigation.cpp:1]: Split your code up in 2, one for audio feedback and one for textual feedback.

For the strings, you could make a lookup table (aka std::map or std::vector), to grab the textual feedback, similarly you can do the same for grabbing the audio files.

**Review by Yimura — CHANGES_REQUESTED** [review]
Where are the instructions for the sound files? Then there's also on how a user would get these files in their navigation folder, how would you handle that?

The feature is nice, however the code could receive some improvements.

File: src/util/audio.hpp (hunk 37 above)
Comment (Rxann) [reply src/util/audio.hpp:145]: I originally created it to allow other people to use audio in an easier fashion. Should I remove it?

File: src/backend/looped/vehicle/navigation.cpp
Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp:1]: So like 2 different looped commands?

File: src/backend/looped/vehicle/navigation.cpp (hunk 5 above)
Comment (Rxann) [reply src/backend/looped/vehicle/navigation.cpp:143]: I am trolling lmao

**Comment by gir489returns** [comment]
I think you should scrap this and try again. There are many different ways to do this much better. In the mission Pacific Standard - Vans, there are 2 teams, driver and navigator. The navigator can issue commands via his arrows to the driver, and they show up on screen as an arrow. I think you should reverse engineer how this script works and abuse it to show the directions of the GPS. You could also abuse the racing cylinders that show a giant yellow circle where you need to turn, with an arrow of the direction for which to turn. There are many more elegant in-engine solutions, but this ain't it chief.

**Comment by Rxann** [comment]
> I think you should scrap this and try again. There are many different ways to do this much better. In the mission Pacific Standard - Vans, there are 2 teams, driver and navigator. The navigator can issue commands via his arrows to the driver, and they show up on screen as an arrow. I think you should reverse engineer how this script works and abuse it to show the directions of the GPS. You could also abuse the racing cylinders that show a giant yellow circle where you need to turn, with an arrow of the direction for which to turn. There are many more elegant in-engine solutions, but this ain't it chief.

Yeah, I can't reverse engineer so that may be a problem. Personally, I did try hard on this but I guess I could scrap it. I have made some progress on making it better though. I guess I will scrap it.

**Comment by Yimura** [comment]
> Yeah, I can't reverse engineer so that may be a problem. Personally, I did try hard on this but I guess I could scrap it. I have made some progress on making it better though. I guess I will scrap it.

The idea was fine really, the execution of it not so much.

From your code in this PR and other PR's on YimMenu what mainly striked me is a lack of code knowledge and structuring of code. It would be best if you spent some time developing some stuff on your own.
Try making a Discord bot, write a webserver from scratch. It doesn't need to be C++, you could use NodeJS, Python, etc...

Anything goes as long as you practice.

**Comment by Rxann** [comment]
> > Yeah, I can't reverse engineer so that may be a problem. Personally, I did try hard on this but I guess I could scrap it. I have made some progress on making it better though. I guess I will scrap it.
> 
> 
> 
> The idea was fine really, the execution of it not so much.
> 
> 
> 
> From your code in this PR and other PR's on YimMenu what mainly striked me is a lack of code knowledge and structuring of code. It would be best if you spent some time developing some stuff on your own.
> 
> Try making a Discord bot, write a webserver from scratch. It doesn't need to be C++, you could use NodeJS, Python, etc...
> 
> 
> 
> Anything goes as long as you practice.
> 
> 

Okay, I mean I have done things in all of these programming languages  before but, before I submit any new pr's I will make sure to give it a solid review and making sure I use things provided by YimMenu especially.

**Comment by ywz978020607** [comment]
hi! I also want to get the purple path info to call third-party voice when I am going to turn left/right, I want to know how you get the path info from game memory? I want to get the path info and send the signal to python or other process to play video.
